### PR TITLE
[FEAT-2496] Add Customize fields button in new column menu

### LINKED ISSUE #2496: Include a button in table views new column menu linking to an object detail page
# Current behavior

CleanShot 2023-11-14 at 11 01 01

Users can show or hide `enabled` fields 

# Additional behavior

We want to add a button that allows users to easily navigate to the object settings page for customizing, editing, and removing fields.

image

### PR DESCRIPTION
**Pull request by umrkhn**
Fixes: #2496

### CODE REVIEW

File: front/src/modules/ui/navigation/menu-item/components/MenuItem.tsx [hunk 1]
```diff
@@ -19,7 +19,7 @@ export type MenuItemProps = {
   LeftIcon?: IconComponent | null;
   accent?: MenuItemAccent;
   text: string;
-  iconButtons?: MenuItemIconButton[];
+  iconButtons?: MenuItemIconButton[] | null;
```
Comment (charlesBochet) [inline front/src/modules/ui/navigation/menu-item/components/MenuItem.tsx:22]: is it working with the current API if we pass an empty array? I feel a null option is duplicated with []

File: front/src/modules/companies/constants/companiesAvailableFieldDefinitions.tsx [hunk 2]
```diff
@@ -179,4 +180,17 @@ export const companiesAvailableFieldDefinitions: ColumnDefinition<FieldMetadata>
       isVisible: false,
       infoTooltipContent: 'The company Twitter account.',
     } satisfies ColumnDefinition<FieldURLMetadata>,
+    {
+      fieldMetadataId: 'customizeFields',
```
Comment (charlesBochet) [inline front/src/modules/companies/constants/companiesAvailableFieldDefinitions.tsx:184]: @umrkhn Let's avoid hacking into the availableFieldDefinitions constants :) What we want here is a hard coded option within RecordTableHeaderPlus button.

We actually have refactored this section and this file doesn't exist anymore anyway as we are computing the available fields dynamically now!

File: front/src/modules/ui/object/record-table/components/RecordTableHeaderPlusButtonContent.tsx [hunk 3]
```diff
@@ -28,19 +30,32 @@ export const RecordTableHeaderPlusButtonContent = () => {
...
         <MenuItem
           key={column.fieldMetadataId}
-          iconButtons={[
-            {
-              Icon: IconPlus,
-              onClick: () => handleAddColumn(column),
-            },
-          ]}
+          iconButtons={
+            column.fieldMetadataId !== 'customizeFields'
```
Comment (charlesBochet) [inline front/src/modules/ui/object/record-table/components/RecordTableHeaderPlusButtonContent.tsx:48]: Let's avoid this and hard-code the option in its own MenuItem below this hiddenTableColumns.map

**Review by charlesBochet — COMMENTED** [review]
@umrkhn Thank you! I have left comment! We have heavily touched this section, you might need to merge main into your branch, or start back from main and re-apply your changes

**Comment by umrkhn** [comment]
> @umrkhn Thank you! I have left comment! We have heavily touched this section, you might need to merge main into your branch, or start back from main and re-apply your changes

I will propose a new solution and discard the previous one as there are a lot of updates, and yes I'll hard-code it like you said.

File: front/src/modules/ui/navigation/menu-item/components/MenuItem.tsx [hunk 4]
```diff
@@ -48,7 +48,7 @@ export const MenuItem = ({
   return (
     <StyledHoverableMenuItemBase
       data-testid={testId ?? undefined}
-      onClick={handleMenuItemClick}
+      onClick={onClick ? handleMenuItemClick : undefined}
```
Comment (charlesBochet) [inline front/src/modules/ui/navigation/menu-item/components/MenuItem.tsx:51]: let's make this change in handleMenuItemClick instead :)

Comment (umrkhn) [reply front/src/modules/ui/navigation/menu-item/components/MenuItem.tsx:51]: sure :+1:

**Comment by charlesBochet** [comment]
@umrkhn The code looks good, I've taken another look at the ticket and the design includes a separator. Could you add it? It should be easy to add, we already have a `DropdownMenuSeparator` component

Another point, we should create link to the field edition page. If we are on a RecordTable of the object "Company" we want to point to `/settings/objects/companies`. To achieve that, you can add the objectMetadataPluralName to the `ObjectMetadataConfig` type. You can write into the objectMetadataConfigState (recoilstate) using:
```
setObjectMetadataConfig?.({
        basePathToShowPage,
        labelIdentifierFieldMetadataId,
      });
```
in RecordTableEffect

Then you can read it in the dropdown using:   `const { objectMetadataConfigState } = useRecordTableScopedStates();`

**Comment by umrkhn** [comment]
> Another point, we should create link to the field edition page. If we are on a RecordTable of the object "Company" we want to point to `/settings/objects/companies`. To achieve that, you can add the objectMetadataPluralName to the `ObjectMetadataConfig` type. You can write into the objectMetadataConfigState (recoilstate) using:
> 
> ```
> setObjectMetadataConfig?.({
>         basePathToShowPage,
>         labelIdentifierFieldMetadataId,
>       });
> ```
> 
> in RecordTableEffect
> 
> Then you can read it in the dropdown using: `const { objectMetadataConfigState } = useRecordTableScopedStates();`

@charlesBochet I have added the DropdownMenuSeparator but I donot understand the second part could you give me more info pls?

**Review by charlesBochet — APPROVED** [review]
@umrkhn We have an ongoing refactoring about metadata. I want to merge this PR first without actually importing objectMetadataItem info. It's fine for now, we will do a v2

Thank you!

**Pull request merged**